akka-stream-tests/src/test/scala/akka/stream/scaladsl/AttributesSpec.scala

```diff
@@ -45,28 +104,138 @@ class AttributesSpec extends StreamSpec {
+
+    "give access to first attribute" in {
+      attributes.getFirst[Name] should ===(Some(Attributes.Name("a")))
+    }
+
+    "give access to attribute byt type" in {
+      attributes.get[Name] should ===(Some(Attributes.Name("b")))
+    }
+
+  }
+
+  "attributes on a stage" must {
+
     "be overridable on a module basis" in {
-      val runnable = Source.empty.toMat(AttributesSink().withAttributes(Attributes.name("new-name")))(Keep.right)
-      whenReady(runnable.run()) { attributes ⇒
-        attributes.get[Name] should contain(Name("new-name"))
-      }
+      val attributes =
+        Source.fromGraph(new AttributesSource().withAttributes(Attributes.name("new-name")))
+          .runWith(Sink.head)
+          .futureValue
+      attributes.get[Name] should contain(Name("new-name"))
     }
 
     "keep the outermost attribute as the least specific" in {
-      val runnable = Source.empty.toMat(AttributesSink())(Keep.right).withAttributes(Attributes.name("new-name"))
-      whenReady(runnable.run()) { attributes ⇒
-        attributes.get[Name] should contain(Name("attributesSink"))
-      }
+      val attributes = Source.fromGraph(new AttributesSource(Attributes.name("original-name")))
+        .map(identity)
+        .addAttributes(Attributes.name("whole-graph"))
+        .runWith(Sink.head)
+        .futureValue
+
+      // most specific
+      attributes.get[Name] should contain(Name("original-name"))
+
+      // least specific
+      attributes.getFirst[Name] should contain(Name("whole-graph"))
     }
 
-    val attributes = Attributes.name("a") and Attributes.name("b") and Attributes.inputBuffer(1, 2)
+    "replace the attributes directly on a graph stage" in {
+      val attributes =
+        Source.fromGraph(
+          new AttributesSource(Attributes.name("original-name")).withAttributes(Attributes.name("new-name"))
+        )
+          .runWith(Sink.head)
+          .futureValue
 
-    "give access to first attribute" in {
-      attributes.getFirst[Name] should ===(Some(Attributes.Name("a")))
+      // most specific
+      attributes.get[Name] should contain(Name("new-name"))
+
+      // least specific
+      attributes.getFirst[Name] should contain(Name("new-name"))
     }
 
-    "give access to attribute byt type" in {
-      attributes.get[Name] should ===(Some(Attributes.Name("b")))
+    "make the attribues on Source.fromGraph source behave the same as the stage itself" in {
+      val attributes =
+        Source.fromGraph(new AttributesSource(Attributes.name("original-name")))
+          .withAttributes(Attributes.name("replaced")) // this actually replaces now
+          .toMat(Sink.head)(Keep.right).withAttributes(Attributes.name("whole-graph"))
+          .run()
+          .futureValue
+
+      // most specific
+      attributes.get[Name] should contain(Name("replaced"))
+
+      // least specific
+      attributes.getFirst[Name] should contain(Name("whole-graph"))
```

Comment:
would be good with better naming of these `.get` methods

Reply:
Those exist nowadays AFAIK

Reply:
That is what I mean, we should deprecate/rename them so that it's possible to understand what they do. Similar with `addAttributes` vs `withAttributes`